File: 50.powx-n.java
class Solution {
    public double myPow(double x, int n) {
        if (n == 0)
            return 1;
        if (n == -1 << 31)
            return 1 / x * this.myPow(x, n + 1);
        if (n < 0)
            return 1 / this.myPow(x, -n);
        double subVal = this.myPow(x, (int) Math.floor(n / 2));
        return n % 2 == 1 ? x * subVal * subVal : subVal * subVal;
    }
}

File: 50.powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 1:
            return x
        if n == 0:
            return 1
        if n < 0:
            return 1 / self.myPow(x, -n)
        ans = 1
        if n % 2 == 1:
            n -= 1
            ans *= x
        v = self.myPow(x, n // 2)
        return v * v * ans
